src/main/java/org/javarosa/core/model/FormDef.java

```diff
@@ -1458,38 +1458,51 @@ public void readExternal(DataInputStream dis, PrototypeFactory pf) throws IOExce
+    }
+
     public void initialize(boolean newInstance, boolean isCompletedInstance,
                            InstanceInitializationFactory factory) {
-        initialize(newInstance, isCompletedInstance, factory, null, false);
+        initialize(newInstance, isCompletedInstance, factory, null, false, true);
     }
 
-    public void initialize(boolean newInstance, InstanceInitializationFactory factory) {
-        initialize(newInstance, false, factory, null, false);
+    public void initialize(boolean newInstance, InstanceInitializationFactory factory, String locale, boolean isReadOnly) {
+        initialize(newInstance, isCompletedInstance, factory, locale, isReadOnly, true);
     }
 
-    public void initialize(boolean newInstance, InstanceInitializationFactory factory, String locale,
-                           boolean isReadOnly) {
-        initialize(newInstance, false, factory, locale, isReadOnly);
+    public void initialize(boolean newInstance, InstanceInitializationFactory factory, String locale) {
+        initialize(newInstance, false, factory, locale, false, true);
     }
 
+    public void initialize(boolean newInstance,
+                           boolean isCompletedInstance,
+                           InstanceInitializationFactory factory,
+                           String locale) {
+        initialize(newInstance, isCompletedInstance, factory, locale, false, true);
+    }
     /**
      * meant to be called after deserialization and initialization of handlers
      *
      * @param newInstance true if the form is to be used for a new entry interaction,
      *                    false if it is using an existing IDataModel
+     * @param isCompletedInstance true if this is an already completed instance we are editing
+     *                            (presumably in HQ) - so don't fire end of form event.
+     * @param factory instance factory containing references to external data sources
      * @param locale The default locale in the current environment, if provided. Can be null
      *               to rely on the form's internal default.
+     * @param isReadOnly true if this form has already been saved and so should not be re-processed
+     *
+     * @param reloadingIncompleteForm true if this form is being reloaded from an incomplete form
+     *                                and so we should re-process triggerables to account for
+     *                                changes to user databases
      */
-    public void initialize(boolean newInstance, boolean isCompletedInstance,
-                           InstanceInitializationFactory factory, String locale, boolean isReadOnly) {
+    public void initialize(boolean newInstance,
+                           boolean isCompletedInstance,
+                           InstanceInitializationFactory factory,
+                           String locale,
+                           boolean isReadOnly,
+                           boolean reloadingIncompleteForm) {
```

Comment:
If I'm reading this correctly, all of the previously-existing versions of `initialize()` now blindly pass in `true` for this argument. But how are you sure that all of those functions are always called only when we're reloading an incomplete form?

Reply:
Yes, that's correct. In Android, this function is only called once when you're opening a new form or re-opening an incomplete form. In Formplayer, we call this function every time you answer a question or interact with the form because we have to serialize and de-serialize and then re-initialize the form each time the server responds. This boolean is a bit of a hack/sneaky way to tell CommCare not to re-initialize triggerables if we're sure the background data hasn't changed - in short, we don't want to call this function for each answer in Formplayer. From the Android perspective I can see this naming doesn't really make sense. I'll update.